### PR TITLE
Expressions: Fix expression load with legacy UID -100

### LINKED ISSUE #64121: From 9.3.8 onwards, existing math expressions are not included in queries
There is very often trouble with the math functions during upgrades, especially between major releases. However, this is a bug in between minor releases. This issue happens when upgrading from 9.3.6 to 9.3.8 and newer. Rolling back to 9.3.6 fixes the issue.

**What happened**:
bilde
The expression is empty. Should include math expression below.

**What you expected to happen**:
bilde

**How to reproduce it (as minimally and precisely as possible)**:
Upgrade from 9.3.6 to any newer release

**Anything else we need to know?**:
Possible to fix the problem by manually entering all math-functions over again. This is tedious.

**Environment**:
- Grafana version: >9.3.8
- Data source type & version:
- OS Grafana is installed on: RHEL9.1
- User OS & Browser: Windows 10 Chrome/Firefox
- Grafana plugins:
- Others:

### PR DESCRIPTION
**Pull request by itsmylife**
**What is this feature?**

When an old dashboard panel has an Expression and its UID is a legacy one (`-100`) panel displays an error and doesn't run the query. This PR is addressing that bug. 

**Who is this feature for?**

All users who have Expression in their panels/dashboards

**Which issue(s) does this PR fix?**:

Fixes https://github.com/grafana/grafana/issues/64121

**Special notes for your reviewer:**
#### How to test?
- Run Grafana with InfluxDB 1.x
- Import dashboard using the JSON below
- Observe the failure
- Switch to this branch and restart Grafana
- Hard refresh the page and observe the panel is working as it should be
<details>

<summary>Test JSON</summary>

```json
{
  "annotations": {
    "list": [
      {
        "builtIn": 1,
        "datasource": {
          "type": "grafana",
          "uid": "-- Grafana --"
        },
        "enable": true,
        "hide": true,
        "iconColor": "rgba(0, 211, 255, 1)",
        "name": "Annotations & Alerts",
        "target": {
          "limit": 100,
          "matchAny": false,
          "tags": [],
          "type": "dashboard"
        },
        "type": "dashboard"
      }
    ]
  },
  "editable": true,
  "fiscalYearStartMonth": 0,
  "graphTooltip": 0,
  "id": 1155,
  "links": [],
  "liveNow": false,
  "panels": [
    {
      "gridPos": {
        "h": 1,
        "w": 24,
        "x": 0,
        "y": 0
      },
      "id": 3,
      "title": "Row title",
      "type": "row"
    },
    {
      "datasource": {
        "type": "influxdb",
        "uid": "${datasource}"
      },
      "fieldConfig": {
        "defaults": {
          "color": {
            "mode": "thresholds"
          },
          "mappings": [],
          "thresholds": {
            "mode": "absolute",
            "steps": [
              {
                "color": "green",
                "value": null
              },
              {
                "color": "red",
                "value": 80
              }
            ]
          }
        },
        "overrides": [
          {
            "matcher": {
              "id": "byName",
              "options": "cpu.mean"
            },
            "properties": [
              {
                "id": "unit",
                "value": "percent"
              },
              {
                "id": "displayName",
                "value": "Attach 3"
              }
            ]
          }
        ]
      },
      "gridPos": {
        "h": 8,
        "w": 12,
        "x": 0,
        "y": 1
      },
      "id": 2,
      "options": {
        "orientation": "auto",
        "reduceOptions": {
          "calcs": [
            "lastNotNull"
          ],
          "fields": "",
          "values": false
        },
        "showThresholdLabels": false,
        "showThresholdMarkers": true
      },
      "pluginVersion": "9.5.0-pre",
      "targets": [
        {
          "datasource": {
            "type": "influxdb",
            "uid": "${datasource}"
          },
          "groupBy": [
            {
              "params": [
                "$__interval"
              ],
              "type": "time"
            },
            {
              "params": [
                "null"
              ],
              "type": "fill"
            }
          ],
          "measurement": "cpu",
          "orderByTime": "ASC",
          "policy": "bar",
          "refId": "A",
          "resultFormat": "time_series",
          "select": [
            [
              {
                "params": [
                  "value"
                ],
                "type": "field"
              },
              {
                "params": [],
                "type": "mean"
              }
            ]
          ],
          "tags": []
        },
        {
          "datasource": {
            "type": "influxdb",
            "uid": "${datasource}"
          },
          "groupBy": [
            {
              "params": [
                "$__interval"
              ],
              "type": "time"
            },
            {
              "params": [
                "null"
              ],
              "type": "fill"
            }
          ],
          "hide": false,
          "measurement": "payment.ended",
          "orderByTime": "ASC",
          "policy": "bar",
          "refId": "C",
          "resultFormat": "time_series",
          "select": [
            [
              {
                "params": [
                  "value"
                ],
                "type": "field"
              },
              {
                "params": [],
                "type": "mean"
              }
            ]
          ],
          "tags": []
        },
        {
          "datasource": {
            "name": "Expression",
            "type": "__expr__",
            "uid": "-100"
          },
          "expression": "100*($A/($C+$A))",
          "hide": false,
          "refId": "X",
          "type": "math"
        },
        {
          "datasource": {
            "name": "Expression",
            "type": "__expr__",
            "uid": "__expr__"
          },
          "expression": "0.5*($A/($C+$A))",
          "hide": false,
          "refId": "Z",
          "type": "math"
        }
      ],
      "title": "Created On Main",
      "type": "gauge"
    }
  ],
  "refresh": "",
  "schemaVersion": 38,
  "style": "dark",
  "tags": [],
  "templating": {
    "list": [
      {
        "current": {
          "selected": false,
          "text": "InfluxDB1-v1.8-8087-InfluxQL",
          "value": "InfluxDB1-v1.8-8087-InfluxQL"
        },
        "hide": 0,
        "includeAll": false,
        "multi": false,
        "name": "datasource",
        "options": [],
        "query": "influxdb",
        "queryValue": "",
        "refresh": 1,
        "regex": "",
        "skipUrlSync": false,
        "type": "datasource"
      }
    ]
  },
  "time": {
    "from": "now-6h",
    "to": "now"
  },
  "timepicker": {},
  "timezone": "",
  "title": "Expressions Investigation",
  "uid": "dc49ec01-c737-4840-b2c3-eb89737b4595",
  "version": 17,
  "weekStart": ""
}
```

</details>

Please check that:
- [x] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: public/app/features/plugins/datasource_srv.ts [hunk 1]
```diff
@@ -65,26 +65,17 @@ export class DatasourceSrv implements DataSourceService {
...
 
     // Expressions has a new UID as __expr__ See: https://github.com/grafana/grafana/pull/62510/
     // But we still have dashboards/panels with old expression UID (-100)
     // To support both UIDs until we migrate them all to new one, this check is necessary
     if (isExpressionReference(nameOrUid)) {
-      return expressionDatasource.instanceSettings;
+      return expressionInstanceSettings;
+    }
+
+    if (nameOrUid === 'default' || nameOrUid === null || nameOrUid === undefined) {
```
Comment (ryantxu) [inline public/app/features/plugins/datasource_srv.ts:77]:
```suggestion
    if (nameOrUid === 'default' || nameOrUid == null) {
```
personal preference that `==` is the same as checking exact equal on both null and undefined.

File: public/app/features/plugins/datasource_srv.ts [hunk 2]
```diff
@@ -114,15 +105,19 @@ export class DatasourceSrv implements DataSourceService {
       };
     }
 
-    return this.settingsMapByUid[nameOrUid] ?? this.settingsMapByName[nameOrUid];
+    return this.settingsMapByUid[nameOrUid] ?? this.settingsMapByName[nameOrUid] ?? this.settingsMapById[nameOrUid];
```
Comment (ryantxu) [inline public/app/features/plugins/datasource_srv.ts:108]: Is there a formatting option that lets us put these on their own lines? 
 it took me a while to read how they are different :)
```
return 
   this.settingsMapByUid[nameOrUid] ??  // UID
   this.settingsMapByName[nameOrUid] ??  // Name
   this.settingsMapById[nameOrUid] // Internal ID
```

Comment (itsmylife) [reply public/app/features/plugins/datasource_srv.ts:108]: I prefer the way you prefer but I do not think that is in our prettier configuration

**Review by bohandley — APPROVED** [review]
LGTM!

**Pull request merged**